### ISSUE TITLE
Masking not applied correctly

### ISSUE THREAD

**Issue by hackelle** [lucidrains/soft-moe-pytorch#10]
I think there is an issue with the masking in the version with dynamic slots: [see code](https://github.com/lucidrains/soft-moe-pytorch/blob/main/soft_moe_pytorch/soft_moe_with_dynamic_slots.py#L137)

Here it says
```
is_padded, x = pad_to_multiple(x, num_experts, dim = -2)

if is_padded:
    if not exists(mask):
        mask = torch.ones(x.shape[:2], device = x.device, dtype = torch.bool)

    _, mask = pad_to_multiple(mask, num_experts, dim = -1, value = False)
```
However, this results in the mask always being `True` if it does not exist because the input is padded before the mask is created; therefore, the second `pad_to_multiple` never changes anything.
In my opinion, one easy workaround would be to pad the input after creating the mask. Although this would calculate the padding twice, it is not compute intense and only requires minimal code changes:
```
is_padded, _ = pad_to_multiple(x, num_experts, dim = -2)

if is_padded:
    if not exists(mask):
        mask = torch.ones(x.shape[:2], device = x.device, dtype = torch.bool)

    _, mask = pad_to_multiple(mask, num_experts, dim = -1, value = False)

is_padded, x = pad_to_multiple(x, num_experts, dim = -2)
```

Please let me know if this is correct of if I misunderstood something.

**Comment by lucidrains** [comment]
@hackelle hey, i think you are correct, thank you for reviewing this

do you want to see if 0.1.9 looks right?

**Comment by hackelle** [comment]
Yes, it looks like it solves the mentioned issue.
